x/warp/contract_warp_handler.go

```diff
@@ -0,0 +1,128 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/subnet-evm/precompile/contract"
+	predicateutils "github.com/ava-labs/subnet-evm/utils/predicate"
+	"github.com/ava-labs/subnet-evm/vmerrs"
+	warpPayload "github.com/ava-labs/subnet-evm/warp/payload"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+)
+
+var (
+	_ messageHandler = addressedPayloadHandler{}
+	_ messageHandler = blockHashHandler{}
+)
+
+var (
+	getVerifiedWarpMessageInvalidOutput   []byte
+	getVerifiedWarpBlockHashInvalidOutput []byte
+)
+
+func init() {
+	res, err := PackGetVerifiedWarpMessageOutput(GetVerifiedWarpMessageOutput{Valid: false})
+	if err != nil {
+		panic(err)
+	}
+	getVerifiedWarpMessageInvalidOutput = res
+
+	res, err = PackGetVerifiedWarpBlockHashOutput(GetVerifiedWarpBlockHashOutput{Valid: false})
+	if err != nil {
+		panic(err)
+	}
+	getVerifiedWarpBlockHashInvalidOutput = res
+}
+
+type messageHandler interface {
+	packFailed() []byte
+	handleMessage(msg *warp.Message) ([]byte, error)
+}
+
+func handleWarpMessage(accessibleState contract.AccessibleState, input []byte, remainingGas uint64, handler messageHandler) ([]byte, uint64, error) {
+	warpIndex, err := UnpackGetVerifiedWarpMessageInput(input)
+	if err != nil {
+		return nil, remainingGas, fmt.Errorf("%w: %s", errInvalidIndexInput, err)
+	}
+	state := accessibleState.GetStateDB()
+	predicateBytes, exists := state.GetPredicateStorageSlots(ContractAddress, warpIndex)
+	predicateResults := accessibleState.GetBlockContext().GetPredicateResults(state.GetTxHash(), ContractAddress)
+	valid := exists && set.BitsFromBytes(predicateResults).Contains(int(warpIndex))
+	if !valid {
+		return handler.packFailed(), remainingGas, nil
+	}
+
+	// Note: we charge for the size of the message during both predicate verification and each time the message is read during
+	// EVM execution because each execution incurs an additional read cost.
+	msgBytesGas, overflow := math.SafeMul(GasCostPerWarpMessageBytes, uint64(len(predicateBytes)))
+	if overflow {
+		return nil, 0, vmerrs.ErrOutOfGas
+	}
+	if remainingGas, err = contract.DeductGas(remainingGas, msgBytesGas); err != nil {
+		return nil, 0, err
```

Comment:
should it be `remainingGas`?

Reply:
We use `0` in all other places where `contract.DeductGas` returns a non-nil error (which is always `vmerrs.ErrOutOfGas`)